programs/bpf_loader/src/syscalls.rs

```diff
@@ -2672,26 +2673,50 @@ where
                 invoke_context,
             )?;
             {
-                callee_account
-                    .set_lamports(*caller_account.lamports)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_data(caller_account.data)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_executable(caller_account.executable)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_owner(caller_account.owner.as_ref())
-                    .map_err(SyscallError::InstructionError)?;
+                if callee_account.get_lamports() != *caller_account.lamports {
+                    callee_account
+                        .set_lamports(*caller_account.lamports)
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+                if callee_account
+                    .is_data_mutable(caller_account.data.len())
+                    .is_ok()
+                    || callee_account.get_data() != caller_account.data
+                {
+                    callee_account
+                        .set_data(caller_account.data)
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                if callee_account.is_executable() != caller_account.executable {
+                    callee_account
+                        .set_executable(caller_account.executable)
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                if callee_account.get_owner() != caller_account.owner {
+                    callee_account
```

Comment:
nit: how about commenting this very nuanced last placement of `.owner` like the other places?: 

```rust
// Change the owner at the end so that we are allowed to change the lamports and data before
```